### PR TITLE
FEATURE: BEND-39 Ip locator package

### PR DESCRIPTION
**Pull request by Kehomaxd**
Implement ip locator

### CODE REVIEW

File: src/http/middleware/middleware.geoIp.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  export const _getGeoIpLocation = (
+    req: express.Request,
+    res: express.Response,
+    next: express.NextFunction
+  ) => {
+    const userIp = requestIp.getClientIp(req);
+    const geoIpLocation = geoip.lookup(userIp);
+
+    //TODO save userIpInfo into request applaying DOM?
+    req.user.userIpInfo = {
```
Comment (whitehatdevv) [inline src/http/middleware/middleware.geoIp.ts:20]: That's impossible if you haven't changed the global object of express.

**Review by whitehatdevv — CHANGES_REQUESTED** [review]

File: src/http/middleware/middleware.geoIp.ts (hunk 1 above)
Comment (OvidioSantoro) [reply src/http/middleware/middleware.geoIp.ts:20]: @whitehatdevv,  we added it into the `TokenInternalInfo` model that is passed to the User as an optional parameter. So it's inside the global object User that is already declared in Express.

File: src/models/types/gen/gen.token.ts [hunk 2]
```diff
@@ -1,10 +1,12 @@
 import {PermissionRoles} from './gen.permissions';
+import {GeoIpLocation} from '../../domain/domain.geoIpLocation';
 
 /**
  * ! Token Payload to save user data
  * * whitehatdevv - 2021/12/14
  */
 export interface TokenInternalInfo {
   userId: string;
+  userIpInfo?: GeoIpLocation;
```
Comment (OvidioSantoro) [inline src/models/types/gen/gen.token.ts:10]: @whitehatdevv, here is where we added it.

File: src/http/middleware/middleware.geoIp.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import * as geoip from 'geoip-lite';
+import requestIp from 'request-ip';
+import express from 'express';
+
+//TODO how to get IP from User
+// https://www.npmjs.com/package/request-ip
+
+//TODO how will our server update the IP DB regularly? (geoip-lite method > updatedb)
```
Comment (OvidioSantoro) [inline src/http/middleware/middleware.geoIp.ts:8]: @whitehatdevv: We have to think how to implement the Geolocation database update and how often. We should do it regularly but it takes some time (between 10 and 20 minutes depending on your machine as we have tested). Maybe once a weekend? More often?

File: src/http/actions/Auth.ts [hunk 4]
```diff
@@ -65,6 +67,7 @@ export module AuthActions {
                   },
                 ],
             ctrl ? dto._id : null
+            //TODO ask manu
```
Comment (Kehomaxd) [inline src/http/actions/Auth.ts:70]: @whitehatdevv  I'm not sure where and how to implement the geoIpLocation middleware in the actions and I'm not sure what this if else does

File: src/models/domain/domain.geoIpLocation.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+ * * kehomaxd - 2022/02/22
+ */
+
+export interface GeoIpLocation {
+  ipAddress: string;
+  city: string;
+  country: string;
+  timezone: string;
+  isVPN: boolean;
+  ll: Location;
```
Comment (OvidioSantoro) [inline src/models/domain/domain.geoIpLocation.ts:14]: @whitehatdevv: An "issue" here. As far as we have seen, the `Login` model doesn't ask for any of these two parameters and they might be legacy requirements, as the ticket is quite old (from before implementing half the database). Should we change the `Login` model or dismiss these requirements?

Comment (OvidioSantoro) [reply src/models/domain/domain.geoIpLocation.ts:14]: Additionally, the `Works` model has a `Latitude` and `Longitude` elements (the equivalent to the ''ll" here). Maybe this was meant for the `Works` model and not `Login`?

File: src/http/actions/Auth.ts (hunk 4 above)
Comment (OvidioSantoro) [reply src/http/actions/Auth.ts:70]: Indeed. We would need a FindOneAndUpdate linked the the success of the Login process, but we're not sure where it should be placed and if we should use the `username` as filtering data for the query or if this `dto._id` is the user id. (I'm not sure either what this line do)

**Review by OvidioSantoro — APPROVED** [review]

**Review by davidcode10 — APPROVED** [review]

**Review by octapf — APPROVED** [review]

**Review by PatyVilas — APPROVED** [review]

**Pull request merged**